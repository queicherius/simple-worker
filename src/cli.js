/* global process */
import commander from 'commander'
import chalk from 'chalk'
import {queueJob, processJob} from './index.js'

const log = {
  info: (string) => console.log(chalk.gray(string)),
  success: (string) => console.log(chalk.green(string)),
  error: (string) => console.log(chalk.bold.red(string))
}

const defaultOptions = {
  validJobs: false
}

export default function (options) {
  options = {...defaultOptions, ...options}
  const args = parseArguments()

  if (options.validJobs) {
    validateJobName(args.name, options.validJobs)
  }

  if (args.queue) {
    return executeQueued(args)
  }

  return executeInline(args)
}

function parseArguments () {
  let jobName = false

  commander
    .usage('job-name [options]')
    .description('Command line interface for executing or queueing jobs')
    .option('-q, --queue', 'Queue the job instead of executing it inline')
    .option('-d, --data [value]', 'The data object for the job, as stringified JSON', (val) => JSON.parse(val), {})
    .action((job) => {
      jobName = job
    })
    .parse(process.argv)

  return {name: jobName, data: commander.data, queue: commander.queue}
}

function validateJobName (name, validNames) {
  if (name && validNames.indexOf(name) !== -1) {
    return true
  }

  validNames = validNames.map(job => `  - ${job}`).join('\n')
  log.error(`This job name is invalid. Please specify one of the following:\n${validNames}`)
  process.exit(1)
}

function executeQueued (args) {
  log.success(`Queueing job "${args.name}"`)
  log.info('Please wait until the queued job gets processed...')

  queueJob({
    name: args.name,
    title: `[CLI] Queued "${args.name}" via the CLI`,
    priority: 'high',
    data: args.data,
    callback: doneOverwrite
  })
}

function executeInline (args) {
  log.success(`Executing job "${args.name}"`)

  const jobOverwrite = {
    log: log.info,
    data: {...args.data, handler: args.name}
  }

  processJob(jobOverwrite, doneOverwrite, true)
}

function doneOverwrite (err, result) {
  if (err) {
    log.error(`An error occurred in the job:\n${err}`)
    return process.exit(1)
  }

  result = result ? `\n${result}` : ''
  log.success(`Job finished successfully.${result}`)
  process.exit(0)
}
