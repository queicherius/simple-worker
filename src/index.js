const BullQueue = require('bull')
const scheduler = require('node-schedule')
const cronstring = require('cronstring')

class SimpleWorker {
  constructor (options) {
    // Validate the setup options
    if (
      !options.name || !options.redis || !options.jobs || !options.logger ||
      !options.logger.info || !options.logger.warn || !options.logger.error
    ) {
      throw new Error('The queue options are not valid.' +
        'Please supply `name`, `redis`, `jobs`, `logger.info`, `logger.warn` & `logger.error`')
    }
    options.jobs.forEach(validateJobConfiguration)

    // Setup the internal values
    this.name = options.name
    this.connection = options.redis
    this.jobConfiguration = toMap(options.jobs, 'name')
    this.logger = options.logger

    this.connect()
  }

  connect () {
    this._queue = new BullQueue(this.name, this.connection)
    this.logger.info('job_queue_connection_ready')

    // Setup queue listeners
    this._queue.on('error', (error) => {
      this.logger.error('job_queue_connection_error', {
        error_message: error.message,
        error_stack: error.stack
      })
    })

    this._queue.on('stalled', (job) => {
      const { jobName, jobData } = splitJobData(job.data)
      this.logger.warn('job_stalled', {
        job_id: job.id,
        attempt: job.attemptsMade,
        name: jobName,
        data: jobData
      })
    })
  }

  add (name, data) {
    const configuration = this.jobConfiguration[name]

    if (!configuration) {
      this.logger.error('job_errored', { name, error_message: 'Job configuration not found' })
      throw new Error(`Job configuration not found`)
    }

    const jobData = Object.assign({}, data || {}, { handler: name })
    const jobOptions = Object.assign(
      { priority: SimpleWorker.PRIORITIES.MEDIUM },
      configuration.options || {},
      { removeOnComplete: true, removeOnFail: true }
    )

    this.logger.info('job_queued', { name, data })
    return this._queue.add('job', jobData, jobOptions)
  }

  schedule () {
    const schedulable = Object.values(this.jobConfiguration).filter(job => job.scheduling)

    schedulable.forEach(job => {
      const schedule = cronstring(job.scheduling) ? cronstring(job.scheduling) : job.scheduling
      this.logger.info('job_scheduled', { name: job.name, schedule })
      scheduler.scheduleJob(schedule, () => this.add(job.name))
    })
  }

  process () {
    this.logger.info('job_processing_ready')

    this._queue.process('job', 1, async (job, done) => {
      const { jobName, jobData } = splitJobData(job.data)
      const configuration = this.jobConfiguration[jobName]

      if (!configuration) {
        return this.logger.error('job_errored', { name: jobName, error_message: 'Job configuration not found' })
      }

      this.logger.info('job_started', {
        job_id: job.id,
        attempt: job.attemptsMade,
        name: jobName,
        data: jobData
      })

      // Add a function so that jobs can log into the same logger
      const loggerWrapper = (level, kind, values) => {
        this.logger[level](kind, {
          job_id: job.id,
          attempt: job.attemptsMade,
          name: jobName,
          data: jobData,
          ...values
        })
      }
      job.info = (kind, values) => loggerWrapper('info', kind, values)
      job.warn = (kind, values) => loggerWrapper('warn', kind, values)
      job.error = (kind, values) => loggerWrapper('error', kind, values)

      // Add a function so that jobs can queue and list jobs
      job.add = (name, data) => this.add(name, data)
      job.list = () => this.list()

      // Setup start date for job duration calculation
      const start = new Date()

      // Execute the handler
      try {
        const result = await promiseTimeout(configuration.handler(job), job.opts.timeout)

        this.logger.info('job_processed', {
          job_id: job.id,
          attempt: job.attemptsMade,
          name: jobName,
          data: jobData,
          duration: new Date() - start,
          result
        })

        done(result)
      } catch (error) {
        if (error instanceof PromiseTimeoutError) {
          this.logger.error('job_timeout', {
            job_id: job.id,
            attempt: job.attemptsMade,
            name: jobName,
            data: jobData,
            duration: new Date() - start
          })
        } else {
          this.logger.error('job_errored', {
            job_id: job.id,
            attempt: job.attemptsMade,
            name: jobName,
            data: jobData,
            duration: new Date() - start,
            error_message: error.message,
            error_stack: error.stack
          })
        }

        done(null, error)
      }
    })
  }

  async list () {
    const jobs = await Promise.all([
      this._queue.getActive(),
      this._queue.getWaiting()
    ])

    return jobs.reduce((arr, x) => arr.concat(x), [])
  }

  pause () {
    return this._queue.pause()
  }

  resume () {
    return this._queue.resume()
  }

  flush () {
    return this._queue.empty()
  }
}

SimpleWorker.PRIORITIES = {
  HIGH: 5,
  MEDIUM: 10,
  LOW: 20
}

function toMap (array, key) {
  let map = {}

  array.forEach(entry => {
    map[entry[key]] = entry
  })

  return map
}

function splitJobData (data) {
  const jobName = data.handler
  const jobData = Object.assign({}, data)
  delete jobData.handler

  return { jobName, jobData }
}

async function promiseTimeout (promise, timeoutMs) {
  if (!timeoutMs) return promise

  let timeoutId
  const timeoutPromise = new Promise((resolve, reject) => {
    timeoutId = setTimeout(() => reject(new PromiseTimeoutError()), timeoutMs)
  })

  return Promise.race([promise, timeoutPromise]).then((result) => {
    clearTimeout(timeoutId)
    return result
  })
}

class PromiseTimeoutError extends Error {
  constructor () {
    super('Promise timed out')
  }
}

function validateJobConfiguration (job) {
  if (!job.name) {
    throw new Error('The job needs a unique name')
  }

  if (!job.handler) {
    throw new Error(`The job ${job.name} needs a handler function`)
  }
}

module.exports = SimpleWorker
