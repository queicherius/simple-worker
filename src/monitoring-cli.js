import chalk from 'chalk'
import sparkly from 'sparkly'
import pad from 'pad-right'
import Duration from 'duration'
import Table from 'cli-table'
import {getData} from './monitoring.js'

const defaultOptions = {
  refreshRate: 2000
}

export default function (options) {
  options = {...defaultOptions, ...options}

  // Setup the monitoring drawing
  setInterval(() => getData().then(drawTable), options.refreshRate)
}

// Draw the monitoring table
function drawTable (jobs) {
  jobs.sort((a, b) => a.name.localeCompare(b.name))

  var table = new Table({
    head: ['Name', 'Queued', 'Active', 'Completed', 'Timeout', 'Failed', 'Processing History'],
    chars: {'mid': '', 'left-mid': '', 'mid-mid': '', 'right-mid': ''},
    style: {head: ['bold']}
  })

  table.push(['', '', '', '', '', '', ''])

  jobs.map(job => {
    table.push([
      job.name,
      job.stats.queued === 0 ? chalk.gray(0) : chalk.magenta(job.stats.queued),
      job.stats.active === 0 ? chalk.gray(0) : chalk.blue(job.stats.active),
      job.stats.completed === 0 ? chalk.gray(0) : chalk.green(job.stats.completed),
      job.stats.timeout === 0 ? chalk.gray(0) : chalk.yellow(job.stats.timeout),
      job.stats.failed === 0 ? chalk.gray(0) : chalk.red(job.stats.failed),
      pad(sparkly(job.history.slice(0, 25).map(x => x[0]), {min: 0}), 30, ' ') +
      job.history.slice(0, 10).map(x => formatStatus(x[1], formatDuration(x[0]))).join(', ')
    ])
  })

  table.push(['', '', '', '', '', '', ''])

  // Calculate the totals
  const totalQueued = jobs.map(job => job.stats.queued).reduce((a, b) => a + b, 0)
  const totalActive = jobs.map(job => job.stats.active).reduce((a, b) => a + b, 0)
  const totalCompleted = jobs.map(job => job.stats.completed).reduce((a, b) => a + b, 0)
  const totalTimeout = jobs.map(job => job.stats.timeout).reduce((a, b) => a + b, 0)
  const totalFailed = jobs.map(job => job.stats.failed).reduce((a, b) => a + b, 0)
  table.push([
    chalk.bold('TOTAL'),
    totalQueued === 0 ? chalk.bold.gray(0) : chalk.bold.magenta(totalQueued),
    totalActive === 0 ? chalk.bold.gray(0) : chalk.bold.blue(totalActive),
    totalCompleted === 0 ? chalk.bold.gray(0) : chalk.bold.green(totalCompleted),
    totalTimeout === 0 ? chalk.bold.gray(0) : chalk.bold.yellow(totalTimeout),
    totalFailed === 0 ? chalk.bold.gray(0) : chalk.bold.red(totalFailed),
    ''
  ])

  // Clear the screen and draw the table
  console.log('\x1B[2J\x1B[0f')
  console.log(table.toString())
}

// Format a job status as a color
function formatStatus (status, string) {
  if (status === 'completed') {
    return chalk.green(string)
  }

  if (status === 'timeout') {
    return chalk.yellow(string)
  }

  return chalk.red(string)
}

// Format a ms duration in human readable format
function formatDuration (ms) {
  var duration = new Duration(new Date(0), new Date(ms))
  return duration.toString(1).split(' ')[0]
}
