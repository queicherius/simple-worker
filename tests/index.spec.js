/* eslint-env node, mocha */
/* global process */
import {expect} from 'chai'
import sleep from 'sleep-promise'
import * as worker from '../src/index'

const listJobs = function (type, state, callback) {
  worker.listJobs(state)
    .then(results => callback(null, results.filter(x => x.data.handler === type)))
    .catch(err => callback(err))
}

function fakeCliProcess (options, args, callback) {
  let tmpArgv = process.argv
  let tmpExit = process.exit
  let tmpLog = console.log

  // Hide output
  console.log = () => false

  // Catch the exit code to give back later
  let exitCode = false
  process.exit = (code = 0) => {
    exitCode = code
  }

  // Fake the CLI arguments
  process.argv = [
    'node',
    '/Users/david/Projects/some-project/src/worker/cli'
  ].concat(args)

  // Execute the cli, clean up after ourselves and return the exit code
  setTimeout(() => {
    worker.cli(options)

    process.argv = tmpArgv
    process.exit = tmpExit
    console.log = tmpLog

    setTimeout(() => {
      callback(exitCode)
    }, 200)
  }, 200)
}

describe('simple-worker', function () {
  this.timeout(5000)

  // Before each test shut down any running processing handlers,
  // start a new queue, wait for the redis connection and then flush queued jobs
  beforeEach(async () => {
    await worker.reset()
  })

  describe('processing', () => {
    it('throws an error if options are missing', () => {
      const error = "A 'name' and a 'title' have to be provided"
      expect(() => worker.queueJob({name: 'asd'})).to.throw(error)
      expect(() => worker.queueJob({title: 'asd'})).to.throw(error)
    })

    it('queues a job with the correct options', (done) => {
      worker.queueJob({name: 'test', title: 'Some test job'})

      listJobs('test', 'inactive', (err, jobs) => {
        expect(err).to.equal(null)
        expect(jobs.length).to.equal(1)

        const job = jobs[0]
        expect(job.data).to.deep.equal({handler: 'test', title: 'Some test job'})
        expect(job._priority).to.equal(-5)
        expect(job._max_attempts).to.equal(1)
        expect(job._backoff).to.deep.equal({delay: 30000, type: 'exponential'})
        expect(parseInt(job._ttl, 10)).to.equal(3600000)
        expect(job._delay).to.be.equal(undefined)
        done()
      })
    })

    it('can list the jobs', async () => {
      await worker.queueJob({name: 'test', title: 'Some test job'})
      await worker.queueJob({name: 'test2', title: 'Some other test job'})
      await worker.queueJob({name: 'test', title: 'Some test job'})

      const jobs = await worker.listJobs('inactive')
      expect(jobs.length).to.equal(3)
      expect(jobs.map(x => x.data.handler)).to.deep.equal(['test', 'test2', 'test'])
    })

    it('can clear the jobs', async () => {
      await worker.queueJob({name: 'test', title: 'Some test job'})
      await worker.queueJob({name: 'test2', title: 'Some other test job'})
      await worker.queueJob({name: 'test', title: 'Some test job'})

      await worker.clearJobs('inactive')
      const jobs = await worker.listJobs('inactive')
      expect(jobs.length).to.equal(0)
    })

    it('schedules a job', (done) => {
      worker.queueJob({
        name: 'test', title: 'A scheduled test job', schedule: '* * * * * *'
      })

      setTimeout(() => {
        listJobs('test', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(3)
          done()
        })
      }, 3500)
    })

    it('processes a job', (done) => {
      let processed = false
      worker.queueJob({name: 'test', title: 'Some test job'})

      worker.registerJob('test', (job, jobDone) => {
        processed = true
        job.log('Hey there')
        jobDone()
      })

      worker.processJobs()

      setTimeout(() => {
        listJobs('test', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(0)
          expect(processed).to.equal(true)
          done()
        })
      }, 500)
    })

    it('handles an error in a synchronous job', (done) => {
      worker.queueJob({name: 'test', title: 'Some test job'})
      worker.queueJob({name: 'test', title: 'Some test job'})

      worker.registerJob('test', () => {
        throw new Error('Oh no.')
      })

      worker.processJobs()

      setTimeout(() => {
        listJobs('test', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(0)
          done()
        })
      }, 500)
    })

    it('handles an error in a promise job', (done) => {
      worker.queueJob({name: 'test', title: 'Some test job'})
      worker.queueJob({name: 'test', title: 'Some test job'})

      worker.registerJob('test', async () => {
        throw new Error('Welp.')
      })

      worker.processJobs()

      setTimeout(() => {
        listJobs('test', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(0)
          done()
        })
      }, 500)
    })
  })

  describe('monitoring', () => {
    it('monitors a queued job', async () => {
      worker.queueJob({name: 'test-queued', title: 'Some test job'})

      const jobs = await worker.monitoring.getData()
      expect(jobs.length).to.equal(1)
      expect(jobs[0].name).to.equal('test-queued')
      expect(jobs[0].stats).to.deep.equal({queued: 1, active: 0, completed: 0, timeout: 0, failed: 0, total: 1})
      expect(jobs[0].history.length).to.equal(0)
    })

    it('monitors an active job', (done) => {
      worker.queueJob({name: 'test-active', title: 'Some test job'})
      worker.registerJob('test-active', (job, jobDone) => {
        setTimeout(() => {
          jobDone()
        }, 3000)
      })
      worker.processJobs()

      setTimeout(async () => {
        const jobs = await worker.monitoring.getData()
        expect(jobs.length).to.equal(1)
        expect(jobs[0].name).to.equal('test-active')
        expect(jobs[0].stats).to.deep.equal({queued: 0, active: 1, completed: 0, timeout: 0, failed: 0, total: 1})
        expect(jobs[0].history.length).to.equal(0)
        done()
      }, 500)
    })

    it('monitors a completed job', (done) => {
      worker.queueJob({name: 'test-completed', title: 'Some test job'})
      worker.registerJob('test-completed', (job, jobDone) => {
        jobDone()
      })
      worker.processJobs()

      setTimeout(async () => {
        const jobs = await worker.monitoring.getData()
        expect(jobs.length).to.equal(1)
        expect(jobs[0].name).to.equal('test-completed')
        expect(jobs[0].stats).to.deep.equal({queued: 0, active: 0, completed: 1, timeout: 0, failed: 0, total: 1})
        expect(jobs[0].history.length).to.equal(1)
        expect(jobs[0].history[0].length).to.equal(3)
        expect(jobs[0].history[0][0]).to.be.a.number
        expect(jobs[0].history[0][1]).to.equal('completed')
        expect(jobs[0].history[0][2]).to.be.a.number
        done()
      }, 500)
    })

    it('monitors a failed job', (done) => {
      worker.queueJob({name: 'test-failed', title: 'Some test job'})
      worker.registerJob('test-failed', () => {
        throw new Error('Oh no.')
      })
      worker.processJobs()

      setTimeout(async () => {
        const jobs = await worker.monitoring.getData()
        expect(jobs.length).to.equal(1)
        expect(jobs[0].name).to.equal('test-failed')
        expect(jobs[0].stats).to.deep.equal({queued: 0, active: 0, completed: 0, timeout: 0, failed: 1, total: 1})
        expect(jobs[0].history.length).to.equal(1)
        expect(jobs[0].history[0].length).to.equal(3)
        expect(jobs[0].history[0][0]).to.be.a.number
        expect(jobs[0].history[0][1]).to.equal('failed')
        expect(jobs[0].history[0][2]).to.be.a.number
        done()
      }, 500)
    })

    it('monitors a timed out job', (done) => {
      worker.queueJob({name: 'test-ttl', title: 'Some test job', ttl: 250})
      worker.registerJob('test-ttl', async (job, jobDone) => {
        await sleep(3000)
        jobDone()
      })
      worker.processJobs()

      setTimeout(async () => {
        const jobs = await worker.monitoring.getData()
        expect(jobs.length).to.equal(1)
        expect(jobs[0].name).to.equal('test-ttl')
        expect(jobs[0].stats).to.deep.equal({queued: 0, active: 0, completed: 0, timeout: 1, failed: 0, total: 1})
        expect(jobs[0].history.length).to.equal(1)
        expect(jobs[0].history[0].length).to.equal(3)
        expect(jobs[0].history[0][0]).to.be.a.number
        expect(jobs[0].history[0][1]).to.equal('timeout')
        expect(jobs[0].history[0][2]).to.be.a.number
        done()
      }, 4000)
    })

    it('monitors retrying jobs correctly', (done) => {
      worker.queueJob({name: 'test-retry', title: 'Some test job', ttl: 250, attempts: 2, backoff: 1000})
      worker.registerJob('test-retry', async (job, jobDone) => {
        await sleep(1000)
        jobDone()
      })
      worker.processJobs()

      setTimeout(async () => {
        const jobs = await worker.monitoring.getData()
        expect(jobs.length).to.equal(1)
        expect(jobs[0].name).to.equal('test-retry')
        expect(jobs[0].stats).to.deep.equal({queued: 0, active: 0, completed: 0, timeout: 2, failed: 0, total: 2})
        expect(jobs[0].history.length).to.equal(2)
        jobs[0].history.map(entry => {
          expect(entry.length).to.equal(3)
          expect(entry[0]).to.be.a.number
          expect(entry[1]).to.equal('timeout')
          expect(entry[2]).to.be.a.number
        })
        done()
      }, 4000)
    })
  })

  describe('web', () => {
    it('can start the interface', () => {
      worker.webInterface(3000)
      worker.webInterface(3001, 'herp', 'derp')
    })
  })

  describe('cli', () => {
    it('processes a job inline', (done) => {
      let processed = false

      worker.registerJob('cli-test-1', (job, jobDone) => {
        processed = true
        job.log('Hey there')
        jobDone()
      })

      fakeCliProcess({}, ['cli-test-1'], exitCode => {
        expect(exitCode).to.equal(0)

        listJobs('cli-test-1', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(0)
          expect(processed).to.equal(true)
          done()
        })
      })
    })

    it('throws an error if the job fails', (done) => {
      let processed = false

      worker.registerJob('cli-test-5', (job) => {
        processed = true
        job.log('Hey there')
        throw new Error('Oh no.')
      })

      fakeCliProcess({}, ['cli-test-5'], exitCode => {
        expect(exitCode).to.equal(1)

        listJobs('cli-test-5', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(0)
          expect(processed).to.equal(true)
          done()
        })
      })
    })

    it('queues a job', (done) => {
      fakeCliProcess({}, ['cli-test-2', '-q'], exitCode => {
        expect(exitCode).to.equal(false)

        listJobs('cli-test-2', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(1)

          const job = jobs[0]
          expect(job.data).to.deep.equal({
            handler: 'cli-test-2',
            title: '[CLI] Queued "cli-test-2" via the CLI'
          })
          expect(job._priority).to.equal(-10)
          expect(job._max_attempts).to.equal(1)
          expect(job._backoff).to.deep.equal({delay: 30000, type: 'exponential'})
          expect(parseInt(job._ttl, 10)).to.equal(3600000)
          expect(job._delay).to.be.equal(undefined)
          done()
        })
      })
    })

    it('can give data to a job', (done) => {
      fakeCliProcess({}, ['cli-test-3', '-q', `--data={"foo": "bar"}`], exitCode => {
        expect(exitCode).to.equal(false)

        listJobs('cli-test-3', 'inactive', (err, jobs) => {
          expect(err).to.equal(null)
          expect(jobs.length).to.equal(1)

          const job = jobs[0]
          expect(job.data).to.deep.equal({
            handler: 'cli-test-3',
            title: '[CLI] Queued "cli-test-3" via the CLI',
            foo: 'bar'
          })
          expect(job._priority).to.equal(-10)
          expect(job._max_attempts).to.equal(1)
          expect(job._backoff).to.deep.equal({delay: 30000, type: 'exponential'})
          expect(parseInt(job._ttl, 10)).to.equal(3600000)
          expect(job._delay).to.be.equal(undefined)
          done()
        })
      })
    })

    it('throws an error if the job name is invalid', (done) => {
      fakeCliProcess({validJobs: ['cli-test-4']}, ['cli-test-99999'], exitCode => {
        expect(exitCode).to.equal(1)
        done()
      })
    })
  })
})
