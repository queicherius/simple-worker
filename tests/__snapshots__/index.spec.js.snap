// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleWorker can add and list jobs from within a job function 1`] = `
Array [
  Object {
    "handler": "testing-processing",
  },
  Object {
    "handler": "testing-processing-2",
  },
]
`;

exports[`SimpleWorker can add and list jobs from within a job function 2`] = `
Array [
  Object {
    "handler": "testing-processing",
  },
  Object {
    "handler": "testing-processing-2",
  },
]
`;

exports[`SimpleWorker can add and list jobs from within a job function 3`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "testing-processing-2",
    },
  },
]
`;

exports[`SimpleWorker can create a new queue instance 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
]
`;

exports[`SimpleWorker can flush the queue 1`] = `
Array [
  Object {
    "handler": "testing-processing",
    "input": "123",
  },
  Object {
    "handler": "testing-processing",
    "input": "456",
  },
  Object {
    "handler": "testing-processing",
    "input": "789",
  },
]
`;

exports[`SimpleWorker can handle an error in the job function (async) 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "testing-errors",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "job_id": "1",
      "name": "testing-errors",
    },
  },
  Object {
    "kind": "job_errored",
    "level": "error",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "duration": 9001,
      "error_message": "The CPU is on fire.",
      "error_stack": "Error: The CPU is on fire.",
      "job_id": "1",
      "name": "testing-errors",
    },
  },
]
`;

exports[`SimpleWorker can handle an error in the job function (sync) 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "testing-errors",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "job_id": "1",
      "name": "testing-errors",
    },
  },
  Object {
    "kind": "job_errored",
    "level": "error",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "duration": 9001,
      "error_message": "The CPU is on fire.",
      "error_stack": "Error: The CPU is on fire.",
      "job_id": "1",
      "name": "testing-errors",
    },
  },
]
`;

exports[`SimpleWorker can pause and resume the processing 1`] = `
Array [
  Object {
    "handler": "testing-processing",
  },
]
`;

exports[`SimpleWorker can pause and resume the processing 2`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_processed",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "duration": 9001,
      "job_id": "1",
      "name": "testing-processing",
      "result": undefined,
    },
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 1`] = `
Array [
  Object {
    "handler": "testing-processing",
    "input": "123",
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 2`] = `
Array [
  Object {
    "handler": "testing-processing",
    "input": "123",
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 3`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": Object {
        "input": "123",
      },
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_function_log_info",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "job_id": "1",
      "name": "testing-processing",
    },
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 4`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": Object {
        "input": "123",
      },
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_function_log_info",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_function_log_warn",
    "level": "warn",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "foo": "bar",
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_function_log_error",
    "level": "error",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_processed",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "duration": 9001,
      "job_id": "1",
      "name": "testing-processing",
      "result": undefined,
    },
  },
]
`;

exports[`SimpleWorker can schedule jobs 1`] = `
Object {
  "handler": "testing-processing",
}
`;

exports[`SimpleWorker can timeout a long running job 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_started",
    "level": "info",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "job_id": "1",
      "name": "testing-processing",
    },
  },
  Object {
    "kind": "job_timeout",
    "level": "error",
    "values": Object {
      "attempt": 0,
      "data": Object {},
      "duration": 9001,
      "job_id": "1",
      "name": "testing-processing",
    },
  },
]
`;

exports[`SimpleWorker errors when the job configuration is missing while adding 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_errored",
    "level": "error",
    "values": Object {
      "error_message": "Job configuration not found",
      "name": "not-existing",
    },
  },
]
`;

exports[`SimpleWorker errors when the job configuration is missing while processing 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queued",
    "level": "info",
    "values": Object {
      "data": undefined,
      "name": "not-existing",
    },
  },
  Object {
    "kind": "job_processing_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_errored",
    "level": "error",
    "values": Object {
      "error_message": "Job configuration not found",
      "name": "not-existing",
    },
  },
]
`;

exports[`SimpleWorker logs internal errors in the queue 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_queue_connection_error",
    "level": "error",
    "values": Object {
      "error_message": "Oh no.",
      "error_stack": "Error: Oh no.",
    },
  },
]
`;

exports[`SimpleWorker logs stalled jobs in the queue 1`] = `
Array [
  Object {
    "kind": "job_queue_connection_ready",
    "level": "info",
    "values": undefined,
  },
  Object {
    "kind": "job_stalled",
    "level": "warn",
    "values": Object {
      "attempt": 1,
      "data": Object {
        "email": "pepe@me.me",
      },
      "job_id": 333,
      "name": "test-handler",
    },
  },
]
`;
